prometheus/reporter.go

```diff
@@ -0,0 +1,254 @@
+package prometheus
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"net/http"
+	"sort"
+	"sync"
+	"time"
+
+	prom "github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/uber-go/tally"
+)
+
+var (
+	errorAlreadyRegistered = fmt.Errorf("Metric already registered")
+)
+
+type metricID string
+
+// Reporter is a prometheus backed tally reporter
+type Reporter struct {
+	counters  map[metricID]*prom.CounterVec
+	gauges    map[metricID]*prom.GaugeVec
+	summaries map[metricID]*prom.SummaryVec
+	mtx       sync.RWMutex
```

Comment:
If you are returning a public struct this is an issue yes.  Here we are returning an interface that scopes the methods quite explicitly.

If the struct and it's access is scoped correctly embedding the mutex and/or waitgroup, etc can make it more terse.  Combining types/using the type system features for more terse expressions is what I would express as "Go-isms".  

E.g. 
type Floats []float64
func (v Floats) Sum() float64 { 
  r := 0
  for _, n := range v {
    r += n 
  }
  return r
}
// use with Floats(someArray).Sum()